pkg/healthchecker/healthchecker.go

```diff
@@ -18,61 +18,152 @@ package healthchecker
 
 import (
 	"context"
-	"database/sql"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+	"syscall"
 	"time"
 
 	"github.com/cenkalti/backoff"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/cockroachdb/cockroach-operator/pkg/kube"
+	"github.com/cockroachdb/cockroach-operator/pkg/resource"
 	"github.com/go-logr/logr"
 	"github.com/pkg/errors"
 	"go.uber.org/zap/zapcore"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+const (
+	underreplicatedmetric = "ranges_underreplicated{store=\"%v\"}"
+	cmdunderreplicted     = "curl -Lk localhost:%s/_status/vars --silent | grep -i '%s'"
 )
 
 type HealthChecker interface { // for testing
-	Probe(ctx context.Context, l logr.Logger, logSuffix string) error
+	Probe(ctx context.Context, l logr.Logger, logSuffix string, partition int) error
 }
 
 type HealthCheckerImpl struct {
-	Db *sql.DB
+	clientset *kubernetes.Clientset
+	scheme    *runtime.Scheme
+	cluster   *resource.Cluster
+	config    *rest.Config
 }
 
-func NewHealthChecker(db *sql.DB) *HealthCheckerImpl {
-	return &HealthCheckerImpl{Db: db}
+func NewHealthChecker(cluster *resource.Cluster, clientset *kubernetes.Clientset, scheme *runtime.Scheme, config *rest.Config) *HealthCheckerImpl {
+	return &HealthCheckerImpl{
+		clientset: clientset,
+		scheme:    scheme,
+		cluster:   cluster,
+		config:    config,
+	}
 }
 
-func (s *HealthCheckerImpl) Probe(ctx context.Context, l logr.Logger, logSuffix string) error {
-	l.V(int(zapcore.DebugLevel)).Info("Health check probe", "label", logSuffix)
-	return s.waitUntilAllReplicasAreRunning(ctx, l, logSuffix)
-}
+func (s *HealthCheckerImpl) Probe(ctx context.Context, l logr.Logger, logSuffix string, nodeID int) error {
```

Comment:
@keith-mcclellan , @udnay , @chrislovecnm please check the logic from the Probe method. Take in consideration that this runs after the rolling update is triggered. The prestop hook already runs when thisProbe method is called so we wait 1 min for the terminationGracePeriodSeconds to end and we start our logic. I commented the sendKillSignal function because we just restarted.

Reply:
So the order here is
1. restart is triggered
2. prestop hook runs calling `cockroach drain node`
3. probe gets called immediately so we sleep 1min?

Reply:
yes